pkg/identity/keystone/token_getter.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package keystone
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	tokens3 "github.com/gophercloud/gophercloud/openstack/identity/v3/tokens"
+
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+// prompt pulls keystone auth url, username and password from a reader
+func prompt(reader io.Reader) (string, string, string, error) {
+	url, err := promptForString("Keystone Auth URL", reader, true)
+	if err != nil {
+		return "", "", "", err
+	}
+
+	user, err := promptForString("Username", reader, true)
+	if err != nil {
+		return "", "", "", err
+	}
+
+	password, err := promptForString("Password", nil, false)
+	if err != nil {
+		return "", "", "", err
+	}
+	return url, user, password, nil
+}
+
+func promptForString(field string, r io.Reader, show bool) (result string, err error) {
+	fmt.Printf("Please enter %s: ", field)
+	if show {
+		_, err = fmt.Fscan(r, &result)
+	} else {
+		var data []byte
+		if terminal.IsTerminal(int(os.Stdin.Fd())) {
+			data, err = terminal.ReadPassword(int(os.Stdin.Fd()))
+			result = string(data)
+		} else {
+			return "", fmt.Errorf("error reading input for %s", field)
+		}
+	}
+	return result, err
+}
+
+// GetToken returns ExecCredential output based on input data from stdin
+// if "intaractive" is set to true, or from env variables otherwise.
+func GetToken(interactive bool) (string, error) {
+	errRespTemplate := `{
+	"apiVersion": "client.authentication.k8s.io/v1alpha1",
+	"kind": "ExecCredential",
+	"spec": {
+		"response": {
+			"code": 401,
+			"header": {},
+		},
+	}
+}`
+
+	respTemplate := `{
+	"apiVersion": "client.authentication.k8s.io/v1alpha1",
+	"kind": "ExecCredential",
+	"status": {
+		"token": "%v",
+		"expirationTimestamp": "%v"
+	}
+}`
+
+	var options gophercloud.AuthOptions
+	var err error
+	if !interactive {
+		options, err = openstack.AuthOptionsFromEnv()
+		if err != nil {
+			msg := fmt.Errorf("Failed to read openstack env vars: %s", err)
+			return "", msg
+		}
+	} else {
+		url, username, password, err := prompt(os.Stdin)
+		if err != nil {
+			return "", err
+		}
+		options = gophercloud.AuthOptions{
+			IdentityEndpoint: url,
+			Username:         username,
+			Password:         password,
+			DomainName:       "default",
```

Comment:
We should not hard coded domain name here.

Also `DomainName` is only applicable for v3. For v2, we use `TenantName`.

Do you mean that only v3 is supported. Or we would support v2 as well.